web.py

```diff
@@ -258,106 +264,36 @@ def preprocess_dataset(trainset_dir, exp_dir, sr, n_p):
-            p = Popen(
-                cmd, shell=True, cwd=now_dir
-            )  # , stdin=PIPE, stdout=PIPE,stderr=PIPE
-            # 煞笔gr, popen read都非得全跑完了再一次性读取, 不用gr就正常读一句输出一句;只能额外弄出一个文本流定时读
-            done = [False]
-            threading.Thread(
-                target=if_done,
-                args=(
-                    done,
-                    p,
-                ),
-            ).start()
-        else:
-            if gpus_rmvpe != "-":
-                gpus_rmvpe = gpus_rmvpe.split("-")
-                leng = len(gpus_rmvpe)
-                ps = []
-                for idx, n_g in enumerate(gpus_rmvpe):
-                    cmd = (
-                        '"%s" infer/modules/train/extract/extract_f0_rmvpe.py %s %s %s "%s/logs/%s" %s '
-                        % (
-                            config.python_cmd,
-                            leng,
-                            idx,
-                            n_g,
-                            now_dir,
-                            exp_dir,
-                            config.is_half,
-                        )
-                    )
-                    logger.info("Execute: " + cmd)
-                    p = Popen(
-                        cmd, shell=True, cwd=now_dir
-                    )  # , shell=True, stdin=PIPE, stdout=PIPE, stderr=PIPE, cwd=now_dir
-                    ps.append(p)
-                # 煞笔gr, popen read都非得全跑完了再一次性读取, 不用gr就正常读一句输出一句;只能额外弄出一个文本流定时读
-                done = [False]
-                threading.Thread(
-                    target=if_done_multi,  #
-                    args=(
-                        done,
-                        ps,
-                    ),
-                ).start()
-            else:
-                cmd = (
-                    config.python_cmd
-                    + ' infer/modules/train/extract/extract_f0_rmvpe_dml.py "%s/logs/%s" '
-                    % (
-                        now_dir,
-                        exp_dir,
-                    )
-                )
-                logger.info("Execute: " + cmd)
-                p = Popen(
-                    cmd, shell=True, cwd=now_dir
-                )  # , shell=True, stdin=PIPE, stdout=PIPE, stderr=PIPE, cwd=now_dir
-                p.wait()
-                done = [True]
-        while 1:
-            with open(
-                "%s/logs/%s/extract_f0_feature.log" % (now_dir, exp_dir), "r"
-            ) as f:
-                yield (f.read())
-            sleep(1)
-            if done[0]:
-                break
-        with open("%s/logs/%s/extract_f0_feature.log" % (now_dir, exp_dir), "r") as f:
-            log = f.read()
+        extracting_thread = threading.Thread(
+            target=call_extract_features,
+            args=(str(log_dir), f0method, config.is_half, config.device),
+        )
+
+        extracting_thread.start()
+        while extracting_thread.is_alive():
+            yield getlog(log_path)
+
+        log = getlog(log_path)
```

Comment:
I think it's not so good. Maybe use pipe/function callback to print the log.

Reply:
I couldn't figure out exactly how you suggested me to do it so i made it extremely hacky :sob: 
```py
        extracting_thread.start()
        with open(log_path, "r") as f:
            while extracting_thread.is_alive():
                lines = f.readlines()
                for line in lines:
                    logger.info(line.strip("\n"))
                
                yield "".join(lines)

            yield f.read()
```

Reply:
I mean not read anything from the log file. That file is just a backup in case the program stopped suddenly and all logs disappeared. You should get the log string from another place ex. pipe or func calls.

Reply:
> I mean not read anything from the log file. That file is just a backup in case the program stopped suddenly and all logs disappeared. You should get the log string from another place ex. pipe or func calls.

So, no reading from the log file at all during execution, just use regular logger instance to print info
understood

Reply:
Implemented in 936f63c